### PR TITLE
Storing Attribute Types in `_type_attributes`

### LINKED ISSUE #835: Store attribute type of `Entry`
**Is your feature request related to a problem? Please describe.**
We want to store the type of all dataclass fields of `Entry` to facilitate the conversion between primitive data in `DataStore` and the actual object [retrieved](https://github.com/asyml/forte/blob/ac34f102ebb3319c1e23f2f81e20a7e9e6c5e60d/forte/data/data_pack.py#L1499)/[set](https://github.com/asyml/forte/blob/ac34f102ebb3319c1e23f2f81e20a7e9e6c5e60d/forte/data/data_pack.py#L1522) by users.

After the integration of `DataStore`, `Entry`'s attributes are stored into `DataStore` object, which only maintains primitive types. When the attribute itself is also an `Entry`, it will be stored as an integer (`tid`) in `DataStore`. When users access this attribute later, it will be converted back to `Entry` again. Currently we determine the type of the attribute value [on the fly](https://github.com/asyml/forte/blob/ac34f102ebb3319c1e23f2f81e20a7e9e6c5e60d/forte/data/data_pack.py#L1514), which is inefficient and unstable.

**Describe the solution you'd like**
We can maintain a dictionary for each dataclass fields that maps the field name to the corresponding type. One potential solution is to use `typing_inspect.get_origin()` and `typing_inspect.get_args()` to parse the typing info stored in `Entry.__dataclass_fields__`.

**Describe alternatives you've considered**
The attribute type info can also be used in `DataStore` to initialize default data structures.

**Additional context**
- This is part of the data efficiency project
- This PR should be made to the `master` branch.
- After the data tuple class is finished, we may switch to a new branch for integration.

### PR DESCRIPTION
**Pull request by Pushkar-Bhuse**
This PR fixes #835. 

### Description of changes
This PR introduces a new structure of `_type_attributes` of Data Store. The `_type_attributes` now store the type values of individual attributes of each entry in a Data Store. For example:
```
DataStore._type_attributes = {
"ft.onto.base_ontology.Document": {
                "attributes": {
                    "document_class": {"index": 4, "type": (list, (str,))},
                    "sentiment": {"index": 5, "type": (dict, (str, float))},
                    "classifications": {
                        "index": 6,
                        "type": (FDict, (str, Classification)),
                    },
                },
                "parent_class": set(),
            },
}
```

### Possible influences of this PR.
Through this PR, `getter` methods of `dataclass` attributes could be more comprehensive and return values based on their respective types. This is particularly useful in the case where attributes of type `Entry` are only stored through their `tid`.  Through this PR, it can be easy to distinguish these attributes from the ones stored as `int`.

### Test Conducted
The tests were conducted on `data_store_test`, `data_pack_test`, `data_store_serialization`

### CODE REVIEW

File: forte/data/ontology/ontology_code_generator.py [hunk 1]
```diff
@@ -1032,16 +1036,17 @@ def parse_entry(
         property_items, property_names = [], []
         for prop_schema in properties:
             # TODO: add test
-            prop_name = prop_schema["name"]
-            if prop_name in RESERVED_ATTRIBUTE_NAMES:
+            prop = (prop_schema["name"], prop_schema["type"])
```
Comment (mylibrar) [inline forte/data/ontology/ontology_code_generator.py:1039]: We might want to document this change somewhere.

File: forte/data/ontology/top.py [hunk 2]
```diff
@@ -1261,6 +1267,15 @@ def __setstate__(self, state):
         self.__dict__.update(state)
         self._modality = getattr(Modality, state["_modality"])
 
+        # During de-serialization, convert the list back to numpy array.
+        if "_embedding" in state:
+            state["_embedding"] = np.array(state["_embedding"])
+        else:
+            state["_embedding"] = np.empty(0)
+
+        if "_cache" in state and isinstance(state["_cache"], list):
```
Comment (mylibrar) [inline forte/data/ontology/top.py:1276]: This checking might not be stable. Can we assume that `_cache` is a numpy array when it was stored as a list in serialization format?

File: forte/data/data_store.py [hunk 3]
```diff
@@ -286,6 +311,20 @@ def __setstate__(self, state):
...
+        # Update `_type_attributes` to store the types of each
+        # entry attribute as well.
+        for tn in self._type_attributes:
+            entry_type = self._add_entry_types(tn)
+            for attr, type_val in entry_type.items():
+                try:
+                    info_dict = self._type_attributes[tn][
+                        constants.TYPE_ATTR_KEY
+                    ][attr]
+                except KeyError:
```
Comment (mylibrar) [inline forte/data/data_store.py:323]: Does this mean that `entry_type` has some new attributes that are not present in `self._type_attributes`?

File: forte/data/data_store.py [hunk 4]
```diff
@@ -537,18 +592,24 @@ def _get_type_info(self, type_name: str) -> Dict[str, Any]:
 
         attr_dict = {}
         attr_idx = constants.ENTRY_TYPE_INDEX + 1
+        type_dict = self._add_entry_types(type_name)
+
         for attr_name in attributes:
-            attr_dict[attr_name] = attr_idx
+            attr_dict[attr_name] = {
```
Comment (mylibrar) [inline forte/data/data_store.py:598]: Let's also update the docstring of `_get_type_info` on this change

File: forte/data/data_store.py [hunk 5]
```diff
@@ -594,16 +655,65 @@ def _default_attributes_for_type(self, type_name: str) -> List:
         attr_dict: Dict = self._get_type_attribute_dict(type_name)
         attr_fields: Dict = self._get_entry_attributes_by_class(type_name)
         attr_list: List = [None] * len(attr_dict)
-        for attr_name, attr_id in attr_dict.items():
+        for attr_name, attr_info in attr_dict.items():
             # TODO: We should keep a record of the attribute class instead of
             # inspecting the class on the fly.
+            attr_id = attr_info[constants.ATTR_INDEX_KEY]
+
             attr_class = get_origin(attr_fields[attr_name].type)
```
Comment (mylibrar) [inline forte/data/data_store.py:663]: Right now we are still calling `self._get_type_attribute_dict(type_name)` and `self._get_entry_attributes_by_class(type_name)`. Can we get these info directly from `self._type_attributes` or `self._get_type_info()`?

File: tests/forte/data/data_store_serialization_test.py [hunk 6]
```diff
@@ -36,19 +39,29 @@ def setUp(self) -> None:
         DataStore._type_attributes = {
             "ft.onto.base_ontology.Document": {
                 "attributes": {
-                    "sentiment": 4,
-                    "classifications": 5,
+                    "document_class": {"index": 4, "type": (list, (str,))},
```
Comment (mylibrar) [inline tests/forte/data/data_store_serialization_test.py:42]: We are adding a new attribute. So will there be an error if we don't include `document_class` here?

File: tests/forte/data/data_store_test.py [hunk 7]
```diff
@@ -126,19 +128,34 @@ def setUp(self) -> None:
         self.reference_type_attributes = {
```
Comment (mylibrar) [inline tests/forte/data/data_store_test.py:128]: Looks like `data_store_test.py` and `data_store_serialization_test.py` share a lot of resources. Maybe we should just keep single source of truth. Note that they do have difference and some of the difference may be intentional.

File: forte/common/constants.py [hunk 8]
```diff
@@ -38,6 +38,14 @@
 # ``_type_attributes`` of ``DataStore``.
 TYPE_ATTR_KEY = "attributes"
 
+# Name of the key to access the type of an attribute from
+# ``_type_attributes`` of ``DataStore``.
+ATTR_TYPE_KEY = "type"
```
[the remark this answers is not in the document]
Comment (Pushkar-Bhuse) [reply forte/common/constants.py:43]: Changing `TYPE_ATTR_KEY` to  `ATTR_INFO_KEY`

File: forte/data/ontology/code_generation_objects.py [hunk 9]
```diff
@@ -864,9 +864,9 @@ def collect_parents(self, node_dict: Dict[str, Set[str]]):
             found_node = search(self.root, search_node_name=node_name)
             if found_node is not None:
                 while found_node.parent.name != "root":
-                    node_dict[
-                        found_node.parent.name
-                    ] = found_node.parent.attributes
+                    node_dict[found_node.parent.name] = set(
```
[the remark this answers is not in the document]
Comment (Pushkar-Bhuse) [reply forte/data/ontology/code_generation_objects.py:867]: That's right

File: forte/data/ontology/top.py (hunk 2 above)
Comment (Pushkar-Bhuse) [reply forte/data/ontology/top.py:1276]: I think we can. Because if `_cache` is not stored as a string (which is in the case of text), it is stored an an `ndarray`. It can be seen by the definition of `set_cache` [here](https://github.com/asyml/forte/blob/8a61272fe2dd7d3f266f4d66a2f0d3e8f6fc3389/forte/data/ontology/top.py#L1226)

File: forte/data/data_store.py (hunk 3 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:323]: Yes

File: forte/data/data_store.py (hunk 5 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:663]: so `self._get_type_attribute_dict(type_name)` actually uses `self._get_type_info()`. With regards to `self._get_entry_attributes_by_class(type_name)`, we need this function because it gives us the `Field` object which us used to fetch the type information.

File: tests/forte/data/data_store_test.py (hunk 7 above)
Comment (Pushkar-Bhuse) [reply tests/forte/data/data_store_test.py:128]: Should we keep that a part of this PR?

Comment (mylibrar) [reply tests/forte/data/data_store_test.py:128]: I think it can be another PR and it's not a prority right now.

File: forte/data/ontology/ontology_code_generator.py [hunk 10]
```diff
@@ -811,22 +812,25 @@ def parse_schema(
...
+                property_name = property[0]
                 # Check if the name is allowed.
                 if not property_name.isidentifier():
                     raise InvalidIdentifierException(
                         f"The property name: {property_name} is not a valid "
                         f"python identifier."
                     )
 
-                if property_name in self.allowed_types_tree[en.class_name]:
+                if property_name in set(
```
[the remark this answers is not in the document]
Comment (Pushkar-Bhuse) [reply forte/data/ontology/ontology_code_generator.py]: We can't do that because the value of `allowed_types_tree` is updated in the loop

File: forte/data/ontology/top.py (hunk 2 above)
Comment (mylibrar) [reply forte/data/ontology/top.py:1276]: Cool. Then let's just add the assumption here in the comments since it might not be true in future when we have a more flexible Payload that can stores any type of data.

File: forte/data/data_store.py (hunk 3 above)
Comment (mylibrar) [reply forte/data/data_store.py:326]: Maybe it's better to add some comments here on why skipping them and when this may happen.

File: forte/data/ontology/ontology_code_generator.py [hunk 11]
```diff
@@ -811,22 +812,25 @@ def parse_schema(
...
+                    val[0] for val in self.allowed_types_tree[en.class_name]
+                ):
                     warnings.warn(
                         f"Attribute type for the entry {en.class_name} "
                         f"and the attribute {property_name} already present in "
                         f"the ontology, will be overridden",
                         DuplicatedAttributesWarning,
                     )
-                self.allowed_types_tree[en.class_name].add(property_name)
+                self.allowed_types_tree[en.class_name].add(tuple(property))
```
Comment (hunterhector) [inline forte/data/ontology/ontology_code_generator.py:833]: Do we need this `tuple` conversion? The type of `property` should already be a `tuple`

File: forte/data/ontology/code_generation_objects.py [hunk 12]
```diff
@@ -797,7 +797,7 @@ def __init__(self, name: str):
         self.children: List[EntryTreeNode] = []
         self.parent: Optional[EntryTreeNode] = None
         self.name: str = name
-        self.attributes: Set[str] = set()
+        self.attributes: Set[Tuple] = set()
```
Comment (hunterhector) [inline forte/data/ontology/code_generation_objects.py:800]: `Tuple` is under specific, what's the structure of this `Tuple`

File: forte/data/ontology/code_generation_objects.py [hunk 13]
```diff
@@ -817,7 +817,7 @@ def add_node(
         self,
         curr_entry_name: str,
         parent_entry_name: str,
-        curr_entry_attr: Set[str],
+        curr_entry_attr: Set[Tuple],
```
Comment (hunterhector) [inline forte/data/ontology/code_generation_objects.py:820]: what tuple

File: forte/data/ontology/code_generation_objects.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/data/ontology/code_generation_objects.py:867]: so where's the note?

File: forte/data/base_pack.py [hunk 14]
```diff
@@ -455,32 +456,53 @@ def on_entry_creation(
             # Use the auto-inferred control component.
             c = self.__control_component
 
-        def entry_getter(cls: Entry, attr_name: str, field_type):
+        def entry_getter(cls: Entry, attr_name: str):
             """A getter function for dataclass fields of entry object.
             When the field contains ``tid``s, we will convert them to entry
```
Comment (hunterhector) [inline forte/data/base_pack.py:461]: This docstring should be more specific on when they will be convert to entries.

File: forte/data/data_store.py [hunk 15]
```diff
@@ -160,9 +161,16 @@ def __init__(
         ``type_name``, their parent entry, and the order of corresponding attributes.
         The keys are fully qualified names of every type; The value is a
         dictionary with two keys. Key ``attribute`` provides an inner dictionary
-        with all valid attributes for this type and the indices of attributes
-        among these lists. Key ``parent_class`` is a string representing the
-        ancestors of this type.
+        with all valid attributes for this type and the information of attributes
```
Comment (hunterhector) [inline forte/data/data_store.py:164]: Can we restructure this docstring as bullet points of numbered lists? A bit hard to follow right now.

File: forte/data/data_store.py [hunk 16]
```diff
@@ -160,9 +161,16 @@ def __init__(
...
+        among these lists. This information is represented as a dictionary. The
+        dictionary has two entries, the first is index which determines the position
+        where an attribute is stored in a data store entry. The second is type, which
+        is a tuple of two elements that provides the type information of a given
+        attribute. The first element is the unsubscripted version of
+        the attribute's type and the second element is the type arguments
```
Comment (hunterhector) [inline forte/data/data_store.py:170]: examples will help this a lot

File: forte/data/data_store.py [hunk 17]
```diff
@@ -174,20 +182,30 @@ def __init__(
...
-            #     "ft.onto.base_ontology.Document": {
-            #       "attributes": {"document_class": 4,
-            #               "sentiment": 5, "classifications": 6},
-            #       "parent_class": set("forte.data.ontology.top.Annotation"), },
-            #     "ft.onto.base_ontology.Sentence": {
-            #       "attributes": {"speaker": 4,
-            #               "part_id": 5, "sentiment": 6,
-            #               "classification": 7, "classifications": 8},
-            #       "parent_class": set(), }
+            #    "ft.onto.base_ontology.Document": {
```
Comment (hunterhector) [inline forte/data/data_store.py:185]: this is a code-block for python, shouldn't comment it (no need to have the #s).

File: forte/data/data_store.py [hunk 18]
```diff
@@ -594,16 +666,90 @@ def _default_attributes_for_type(self, type_name: str) -> List:
...
+            attributes: This argument is used when parsing ontology
+                files. The entries in the set are a tuples of two
+                elements.
+
+                .. code-block:: python
+
+                    attributes =  {
+                                ('passage_id', 'str'),
+                                ('author', 'str')
+                            }
```
Comment (hunterhector) [inline forte/data/data_store.py:715]: where's return?

File: forte/data/data_store.py [hunk 19]
```diff
@@ -594,16 +666,90 @@ def _default_attributes_for_type(self, type_name: str) -> List:
...
             if attr_class in (FList, list, List):
                 attr_list[attr_id - constants.ATTR_BEGIN_INDEX] = []
             elif attr_class in (FDict, dict, Dict):
                 attr_list[attr_id - constants.ATTR_BEGIN_INDEX] = {}
         return attr_list
 
+    def _new_entry_types(
+        self, type_name: str, attributes: Optional[Set[Tuple[str, str]]] = None
+    ) -> Dict[str, Tuple]:
+        r"""This function takes a fully qualified ``type_name`` class name
```
Comment (hunterhector) [inline forte/data/data_store.py:684]: This function takes two items.

File: forte/data/data_store.py [hunk 20]
```diff
@@ -594,16 +666,90 @@ def _default_attributes_for_type(self, type_name: str) -> List:
...
+        of two elements. The first element is the unsubscripted version of
+        the attribute's type and the second element is the type arguments
+        for the same. The `type_dict` is used to populate the type
+        information for attributes of an entry specified by ``type_name``
+        in `_type_attributes`.
+
+
+        Args:
+            type_name: A fully qualified name of an entry class.
+            attributes: This argument is used when parsing ontology
```
Comment (hunterhector) [inline forte/data/data_store.py:706]: Can't understand this docstring. When should we provide this?

File: forte/data/data_store.py [hunk 21]
```diff
@@ -594,16 +666,90 @@ def _default_attributes_for_type(self, type_name: str) -> List:
...
+                attr_class = get_origin(attr_info.type)
+                attr_args = get_args(attr_info.type)
+
+                # Prior to Python 3.7, typing.List and typing.Dict
+                # is not converted to primitive forms of list and
+                # dict. We handle them separately here
+                if attr_class == Dict:
+                    attr_class = dict
+                if attr_class == List:
+                    attr_class = list
```
Comment (hunterhector) [inline forte/data/data_store.py:747]: I am not confident in this function. Are we only handling these types here? What if we are dealing with something new, like `Set`? How can we make sure this function is robust?

File: forte/data/data_store.py [hunk 22]
```diff
@@ -690,6 +836,34 @@ def _is_annotation(self, type_name: str) -> bool:
             for entry_class in (Annotation, AudioAnnotation)
         )
 
+    def get_entry_types(
```
Comment (hunterhector) [inline forte/data/data_store.py:839]: The function name does not match with its behavior

File: forte/data/data_store.py [hunk 23]
```diff
@@ -594,16 +666,90 @@ def _default_attributes_for_type(self, type_name: str) -> List:
...
+            attr_id = attr_info[constants.ATTR_INDEX_KEY]
+
             attr_class = get_origin(attr_fields[attr_name].type)
             if attr_class in (FList, list, List):
                 attr_list[attr_id - constants.ATTR_BEGIN_INDEX] = []
             elif attr_class in (FDict, dict, Dict):
                 attr_list[attr_id - constants.ATTR_BEGIN_INDEX] = {}
         return attr_list
 
+    def _new_entry_types(
```
Comment (hunterhector) [inline forte/data/data_store.py:681]: what is `new` here in the function name?

File: forte/data/base_pack.py [hunk 24]
```diff
@@ -455,32 +456,53 @@ def on_entry_creation(
...
             """
+
             data_store_ref = (
                 cls.pack._data_store  # pylint: disable=protected-access
             )
             attr_val = data_store_ref.get_attribute(
                 tid=cls.tid, attr_name=attr_name
             )
-            if field_type in (FList, FDict):
+            entry_type = data_store_ref.get_entry_types(
```
Comment (hunterhector) [inline forte/data/base_pack.py:475]: It is not an `entry_type`, it is the type of the attribute. Variable names also affect the quality of the code.

File: forte/data/base_pack.py [hunk 25]
```diff
@@ -455,32 +456,53 @@ def on_entry_creation(
...
+                cls.entry_type(), attr_name
+            )
+
+            if entry_type[0] in (FList, FDict):
```
Comment (hunterhector) [inline forte/data/base_pack.py:479]: not sure if hardcoding `FList` and `FDict` is a good idea. 

We have some constants here. https://github.com/asyml/forte/blob/master/forte/data/ontology/ontology_code_const.py#L86

File: forte/data/data_store.py [hunk 26]
```diff
@@ -258,18 +276,28 @@ def __getstate__(self):
         """
         state = super().__getstate__()
         state["_DataStore__elements"] = {}
+
+        # Make a copy of the updated type_attributes
+        type_attributes = deepcopy(DataStore._type_attributes)
```
Comment (hunterhector) [inline forte/data/data_store.py:281]: why do we need a deepcopy?

**Comment by hunterhector** [comment]
In the future, for this type of more complex PR's , the better idea is to split out small steps and tackle them each one separately.

File: forte/data/data_store.py (hunk 16 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:170]: The example is given below

File: forte/data/data_store.py (hunk 26 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:281]: It is because we want to remove only the type information from the type_attributes dict before serialization. However, since a dict is mutable in python, we cannot do this operation on the original `_type_attributes` dict. Thus, we create a copy of this dictionary, remove the type information from that and then use it for serialization. (Here, type information mainly means the dictionary stored by key `type` in the `_type_attributes` for each entry)

File: forte/data/data_store.py (hunk 23 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:681]: This is function is invoked when new entries are added to `_type_attributes` and thus their type information will also need to be added.

File: forte/data/ontology/code_generation_objects.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Pushkar-Bhuse) [reply forte/data/ontology/code_generation_objects.py:867]: I'll add it to the docstrings/comments of the individual variables.

File: forte/data/ontology/top.py [hunk 27]
```diff
@@ -1250,6 +1250,12 @@ def __getstate__(self):
         # Entry store is being integrated into DataStore
         state = self.__dict__.copy()
         state["_modality"] = self._modality.name
+
+        if isinstance(state["_cache"], np.ndarray):
+            state["_cache"] = list(self._cache.tolist())
+        if isinstance(state["_embedding"], np.ndarray):
+            state["_embedding"] = list(self._embedding.tolist())
```
[the remark this answers is not in the document]
Comment (Pushkar-Bhuse) [reply forte/data/ontology/top.py:1257]: Because they were numpy arrays, they were causing errors during serialization. That is why we added this check.

File: tests/forte/data/data_store_serialization_test.py [hunk 28]
```diff
@@ -73,39 +86,39 @@ def setUp(self) -> None:
...
+                        {"Negative": 0},
                     ],
                     [
                         10,
                         25,
                         3456,
                         "ft.onto.base_ontology.Document",
                         "Negative",
-                        "Class B",
+                        {},
```
[the remark this answers is not in the document]
Comment (Pushkar-Bhuse) [reply tests/forte/data/data_store_serialization_test.py:97]: I was thinking of keeping the data store entry structure consistent with the type information provided in `_type_attributes`. But I can change it back to the original form.

File: forte/data/ontology/top.py [hunk 29]
```diff
@@ -1261,6 +1267,20 @@ def __setstate__(self, state):
         self.__dict__.update(state)
         self._modality = getattr(Modality, state["_modality"])
 
+        # During de-serialization, convert the list back to numpy array.
+        if "_embedding" in state:
+            state["_embedding"] = np.array(state["_embedding"])
+        else:
+            state["_embedding"] = np.empty(0)
```
[the remark this answers is not in the document]
Comment (Pushkar-Bhuse) [reply forte/data/ontology/top.py:1274]: `_embedding` was being stored an an `np.array` which was causing problems during serialization.

File: forte/data/data_store.py (hunk 21 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:747]: So this is an issue caused by the change in the typing module from python 3.6 to python 3.7. Another workaround for this is that we can let the the output `get_origin` be stored in any way depending on the python version and just handle this in the test cases. This is similar to a suggestion made [here](https://github.com/ilevkivskyi/typing_inspect/issues/27)

File: forte/data/base_pack.py (hunk 25 above)
Comment (Pushkar-Bhuse) [reply forte/data/base_pack.py:479]: Can I add constants here?

File: forte/data/data_store.py (hunk 26 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:281]: It is because we want to only remove the type information atom the `_type_attributes` before serialization without affecting the class variable. Since `dict` is mutable, we create a copy of `_type_attributes` which will be modified and then used during serialization.

File: forte/data/base_pack.py (hunk 25 above)
Comment (Pushkar-Bhuse) [reply forte/data/base_pack.py]: I'm not sure if we can use constants from here in base_pack because it makes a partial import of `PackType` from base_pack itself.

Comment (Pushkar-Bhuse) [reply forte/data/base_pack.py]: I have instead added a constant in `core.py` where these classes are defined.

File: forte/data/data_store.py (hunk 16 above)
Comment (hunterhector) [reply forte/data/data_store.py]: it is pretty hard to understand this section now

File: forte/data/data_store.py (hunk 21 above)
Comment (hunterhector) [reply forte/data/data_store.py]: But how can we make sure we we list all these cases, how about `Set`?

File: forte/data/data_store.py (hunk 23 above)
Comment (hunterhector) [reply forte/data/data_store.py:690]: 1. Why is type_attribute information new when an entry is added? Some entries could have the same type
2. Still, I don't think anything is "new", the information is stored and get via `self._get_entry_attributes_by_class(type_name)`, so this function only `get` this information

File: forte/data/base_pack.py (hunk 25 above)
Comment (hunterhector) [reply forte/data/base_pack.py]: they should be better in the same place and nearby, probably here: https://github.com/asyml/forte/blob/master/forte/data/ontology/ontology_code_const.py#L86

Comment (Pushkar-Bhuse) [reply forte/data/base_pack.py]: As I mentioned above, if we put it there we won't be able to use it in `base_pack`

File: forte/data/data_store.py (hunk 23 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:690]: So this function is invoked in the case where it is the first time an entry of this type is being created (and thus it is not present in `_type_attributes`. But I do get your point, I think a better name for this function could be `fetch_entry_type_data`. Would this be okay?

File: forte/data/data_store.py (hunk 16 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py]: Oh, I'll try explaining it better. Is there any area specifically that you feel needs to change?

File: forte/data/data_store.py (hunk 21 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py]: I have added a more general solution in the updated version of this PR which does not require modifying test cases

File: forte/data/data_store.py (hunk 23 above)
Comment (hunterhector) [reply forte/data/data_store.py:690]: let's use the new name, since the "add" action seems to happen after this.

File: forte/data/base_pack.py (hunk 25 above)
Comment (hunterhector) [reply forte/data/base_pack.py]: huh, let's see whether this workout.

File: forte/data/data_store.py (hunk 26 above)
Comment (hunterhector) [reply forte/data/data_store.py:290]: so this variable will go to serialized data. I normally prefer the other way around where we assemble the serialized data instead of removing stuff from a dict. One reason is that we have more contorl over the data being serialized

Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:290]: Yes I actually had started off with your approach, but I noticed that it made the code a bit large unnecessarily because the part to remove (type information) is much lesser than the part to keep. But if you want I can try the other way round as well.

File: forte/data/data_store.py (hunk 23 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:690]: Sounds good

File: forte/data/data_store.py (hunk 26 above)
Comment (hunterhector) [reply forte/data/data_store.py:290]: ok, btw. is it necessary to remove the type information?

Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:290]: Yes, because they are not stored in serializable formats

File: forte/data/ontology/code_generation_objects.py [hunk 30]
```diff
@@ -856,17 +856,19 @@ def collect_parents(self, node_dict: Dict[str, Set[str]]):
 
         Args:
             node_dict: the nodes dictionary of nodes to collect parent nodes
-                for.
+                for. The entry represented by nodes in this dictionary do not store
```
Comment (mylibrar) [inline forte/data/ontology/code_generation_objects.py:859]: The entry node actually stores the type info of attributes according to your latest changes.

File: forte/data/data_store.py [hunk 31]
```diff
@@ -157,12 +164,22 @@ def __init__(
...
         """
         The ``_type_attributes`` is a private dictionary that provides
-        ``type_name``, their parent entry, and the order of corresponding attributes.
-        The keys are fully qualified names of every type; The value is a
-        dictionary with two keys. Key ``attribute`` provides an inner dictionary
-        with all valid attributes for this type and the indices of attributes
-        among these lists. Key ``parent_class`` is a string representing the
-        ancestors of this type.
+        ``type_name``, their parent entry, and the metadata of its corresponding
+        attributes. This metadata includes the order and type information of attributes
```
Comment (mylibrar) [inline forte/data/data_store.py:168]: Feels like it would be better to introduce key and value before explaining the metadata of attributes.

File: forte/data/ontology/code_generation_objects.py (hunk 30 above)
Comment (Pushkar-Bhuse) [reply forte/data/ontology/code_generation_objects.py:859]: So `node_dict` is another component just used in `collect_parents`, this does not store type information because we do [this](https://github.com/Pushkar-Bhuse/forte/blob/39b4de96d625ea3231b35432847bc2f467b27aaa/forte/data/ontology/code_generation_objects.py#L870). On the other hand, [here](https://github.com/Pushkar-Bhuse/forte/blob/39b4de96d625ea3231b35432847bc2f467b27aaa/forte/data/data_store.py#L1852) we use the node with its index and type information and I've mentioned it in the comments.

File: forte/data/data_store.py (hunk 31 above)
Comment (Pushkar-Bhuse) [reply forte/data/data_store.py:168]: So key and value here refer to the components of Python `dict`. Do you mean mentioning that they key-value refer to that of a Python `dict`?

Comment (mylibrar) [reply forte/data/data_store.py:168]: I'm just thinking moving this sentence `The keys are fully qualified ...` before `This metadata includes the order ...`

Comment (Pushkar-Bhuse) [reply forte/data/data_store.py]: Got it

Comment (mylibrar) [reply forte/data/data_store.py]: The docstring still looks a little bit off.

> The ``_type_attributes`` is a private dictionary that provides ``type_name``, ie. their parent entry (represented by the fully qualified name of the entry) as the key, and the

The keys are not parent_entry. And ``type_name`` is different from ``their parent entry``.

Comment (Pushkar-Bhuse) [reply forte/data/data_store.py]: How about - "The ``_type_attributes`` is a private dictionary that contains the parent entry ( (represented by the fully qualified name of the entry) as key and the metadata ...." ?

Comment (mylibrar) [reply forte/data/data_store.py]: I still don't think `parent entry` is a key of `_type_attributes`. `parent_class` may be a key of `_type_attributes[type_name]` but it's not a key for `_type_attributes`.

**Review by mylibrar — APPROVED** [review]
